### PR TITLE
Add test cases for updating sagas

### PR DESCRIPTION
**Pull request by weralabaj**
This PR includes more test cases to `Persistence Backwards Compatibility` suite

### CODE REVIEW

**Comment by tmasternak** [comment]
ping @Particular/endtoend-maintainers

File: src/NHPersistenceCompatibilityTests/4.5/NHibernateSessionFactory.cs [hunk 1]
```diff
@@ -1,19 +1,15 @@
 ﻿using Common;
 using DataDefinitions;
-using NHibernate;
 using NHibernate.Tool.hbm2ddl;
 using NServiceBus.SagaPersisters.NHibernate.AutoPersistence;
 
 namespace Version_4_5
 {
+    using NHibernate;
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/4.5/NHibernateSessionFactory.cs]: why mixing inner and outer usings ?

File: src/NHPersistenceCompatibilityTests/5.0/5.0.csproj [hunk 2]
```diff
@@ -17,7 +17,7 @@
     <DebugType>full</DebugType>
     <Optimize>false</Optimize>
     <OutputPath>..\bin\Debug\5.0\</OutputPath>
-    <DefineConstants>DEBUG;TRACE</DefineConstants>
+    <DefineConstants>TRACE;DEBUG;VERSION_5_0</DefineConstants>
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/5.0/5.0.csproj]: so the VERSION_5_0 constant should only be in debug build?

File: src/NHPersistenceCompatibilityTests/7.0/NHibernateSessionFactory.cs [hunk 3]
```diff
@@ -31,7 +26,9 @@ static ISessionFactory Init()
 
             new SchemaUpdate(configuration).Execute(false, true);
 
-            return configuration.BuildSessionFactory();
+            var sessionFactory = configuration.BuildSessionFactory();
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/7.0/NHibernateSessionFactory.cs]: just return. no need for a local

File: src/NHPersistenceCompatibilityTests/7.0/Persister.cs [hunk 4]
```diff
@@ -7,7 +7,7 @@
 
 public class Persister
 {
-    readonly SagaPersister persister;
+    readonly ISagaPersister persister;
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/7.0/Persister.cs]: kill the readonly

File: src/NHPersistenceCompatibilityTests/4.5/NHibernateSessionFactory.cs (hunk 1 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/4.5/NHibernateSessionFactory.cs]: moved to outer scope

File: src/NHPersistenceCompatibilityTests/5.0/5.0.csproj (hunk 2 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/5.0/5.0.csproj]: no my bad. Fixed

File: src/NHPersistenceCompatibilityTests/7.0/NHibernateSessionFactory.cs (hunk 3 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/7.0/NHibernateSessionFactory.cs]: done

File: src/NHPersistenceCompatibilityTests/7.0/Persister.cs (hunk 4 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/7.0/Persister.cs]: all readonly modifiers removed

**Comment by tmasternak** [comment]
@Particular/sqlserver-transport-maintainers this is green. Ready to merge.

File: src/NHPersistenceCompatibilityTests/SharedDefinitions/ObjectFetcher.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+using System.Collections.Generic;
+
+
+static class ObjectFetcher
+{
+    public static void Traverse(object instance, Type instanceType)
+    {
+        foreach (var propertyInfo in instanceType.GetProperties())
+        {
+            if (ReferenceEquals(propertyInfo.DeclaringType, typeof(object)))
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/SharedDefinitions/ObjectFetcher.cs]: I am surprised referenceequals even works in this context

File: src/NHPersistenceCompatibilityTests/SharedDefinitions/ObjectFetcher.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+            {
+                continue;
+            }
+
+            if (propertyInfo.PropertyType.GetInterface(typeof(IEnumerable<>).FullName) != null)
+            {
+                propertyInfo.GetValue(instance, null)?.ToString();
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/SharedDefinitions/ObjectFetcher.cs]: Why tostring?

File: src/NHPersistenceCompatibilityTests/SharedDefinitions/RawPersister.cs [hunk 7]
```diff
@@ -8,25 +9,54 @@ public void Save(string typeFullName, string body, string correlationPropertyNam
...
-        var getMethod = persister.GetType().GetMethod(nameof(Get)).MakeGenericMethod(sagaDataType);
+        updateMethod.Invoke(persister, new[] { sagaData });
+    }
+
+    public object Get(string typeFullName, Guid id)
+    {
+        var sagaDataType = Type.GetType(typeFullName);
+
+        var persister = CreatePersisterInstance();
+        var getMethod = GetMethod(persister, nameof(Get), sagaDataType);
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/SharedDefinitions/RawPersister.cs]: Isnt nameof get wrong in this context? Isnt it ma coincidence that currenjt methid name matched thename of the method u r returning

File: src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/Util/PersisterFacade.cs [hunk 8]
```diff
@@ -31,5 +38,17 @@ public T Get<T>(Guid sagaId)
...
+
+        public T GetByCorrelationId<T>(string correlationPropertyName, object correlationPropertyValue)
+        {
+            var body = string.Empty;
+            var sagaDataType = typeof(T);
+
+            _rawPersister.Run(p => body = (string)p.GetByCorrelationProperty(sagaDataType.FullName, correlationPropertyName, correlationPropertyValue));
+
+            var result = JsonConvert.DeserializeObject(body, sagaDataType);
+
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/Util/PersisterFacade.cs]: Isnt there a generic api for deserializing?

File: src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/Util/PersisterFacade.cs [hunk 9]
```diff
@@ -20,6 +20,13 @@ public void Save<T>(T instance, string correlationPropertyName, string correlati
             _rawPersister.Run(p => p.Save(instance.GetType().FullName, body, correlationPropertyName, correlationPropertyValue));
         }
 
+        public void Update<T>(T instance)
+        {
+            var body = JsonConvert.SerializeObject(instance);
+
+            _rawPersister.Run(p => p.Update(instance.GetType().FullName, body));
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/Util/PersisterFacade.cs]: Can we remove the underscore

File: src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/UpdatingSagaTests.cs [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+
+            sourcePersister.Save(writeData, nameof(writeData.Id), writeData.Id.ToString());
+
+            writeData.Composite.Value = "updated";
+            sourcePersister.Update(writeData);
+
+            var readData = destinationPersister.Get<TestSagaDataWithComposite>(writeData.Id);
+
+            Assert.AreEqual(writeData.Id, readData.Id);
+            CollectionAssert.AreEqual("updated", readData.Composite.Value);
```
Comment (SimonCropp) [inline src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/UpdatingSagaTests.cs]: Why a collection assert on a string

File: src/NHPersistenceCompatibilityTests/SharedDefinitions/ObjectFetcher.cs (hunk 6 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/SharedDefinitions/ObjectFetcher.cs]: It enforces fetching when property is lazy-initialized proxy

File: src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/Util/PersisterFacade.cs (hunk 9 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/Util/PersisterFacade.cs]: done

File: src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/UpdatingSagaTests.cs (hunk 10 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/UpdatingSagaTests.cs]: Mistake. Fixed.

File: src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/Util/PersisterFacade.cs (hunk 8 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/PersistenceCompatibilityTests/Util/PersisterFacade.cs]: Switched to generic.

File: src/NHPersistenceCompatibilityTests/SharedDefinitions/ObjectFetcher.cs (hunk 5 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/SharedDefinitions/ObjectFetcher.cs]: `A Type object that represents a type is unique; that is, two Type object references refer to the same object if and only if they represent the same type`. More details [here] (https://msdn.microsoft.com/en-us/library/system.type(v=vs.110).aspx) in `Comparing type objects for equality` section

File: src/NHPersistenceCompatibilityTests/SharedDefinitions/RawPersister.cs (hunk 7 above)
Comment (tmasternak) [inline src/NHPersistenceCompatibilityTests/SharedDefinitions/RawPersister.cs]: I was looking at that more like on convention. `Persister`s from different versions should implement methods which are named the same as those of `RawPersister`. 

I've switched to strings to make it make sure it's more visible that names come from `Persister` types.

**Pull request merged**